dev/ese/published/inc/sync.hxx

```diff
@@ -2017,129 +2009,33 @@ class CSemaphoreState
 
         //  data members
 
-        //    transacted state representation (switched on bit 31)
+        //    transacted state representation
 
         union
         {
-            //  Mode 0:  no waiters
-
-            volatile LONG       m_cAvail;       //  0 <= m_cAvail <= ( 1 << 31 ) - 1
-
-            //  Mode 1:  waiters
+            volatile QWORD      m_qwState;
```

Comment:
It should be possible to keep the state as a LONG in order to continue supporting 32-bit, e.g. ARM.

Reply:
Could you please elaborate on the "continue supporting 32-bit" part? Does it refer to the lack of support for certain 64-bit atomics on some platforms? Or is the incompatibility caused by changing the size of the state from 4 to 8 bytes under 32 bits?

The former doesn't look like a problem, at least based on what I could grasp from the code. I see that various parts of ESE rely on the presence of such atomics — for example in the implementation of the [`CInvasiveConcurrentModSet<>`](https://github.com/microsoft/Extensible-Storage-Engine/blob/3bcf428c8a041733043e18fd9ae55cffeba307b2/dev/ese/published/inc/collection.hxx#L1099) or in [`LOG_WRITE_BUFFER::ErrLGScheduleWrite()`](https://github.com/microsoft/Extensible-Storage-Engine/blob/3bcf428c8a041733043e18fd9ae55cffeba307b2/dev/ese/src/ese/_log/logwrite.cxx#L2312).

If it's the latter, then the "continue supporting" part is still somewhat unclear to me. I think the `CSemaphore` does not participate in the on-disk format and is not a part of the ABI. So I was hoping that this change doesn't do anything beyond increasing the in-memory size of certain structures under 32 bits.

In general, I don't yet see an acceptable approach where we get to keep the state in 32 bits. With `WaitOnAddress()` we require a value for the semaphore's count to wait on, but we also have to keep a separate record of the number of waiting threads. So unless there is a trick that I haven't thought of, I see two options:

1. Make those two values 32-bit, forming a 64-bit state
2. Make those two values 16-bit, forming a 32-bit state

Option 1. is what the patch does.

Option 2. doesn't seem to be practically feasible, because the `CSemaphore` currently accepts an `INT` as its semaphore count.  Switching that to 16 bits would narrow the actual range of the accepted values, and I think that would constitute an unsafe change that requires validating and updating every usage of the semaphore.